### PR TITLE
[RM] Allows custom config for inline delimiters

### PR DESCRIPTION
**Pull request by rachelthompson**
These changes allow the consuming application to pass in a custom
configuration for parsing Remark-Math segments inline.

The intended use case is that delimiters can indicate either Inline Mode
or Display (Block) Mode, depending on the delimiter used, and custom
delimiters for either of these.

For the particular use case we are applying, we don't want
single-dollar-sign delimited strings to be interpreted as LaTeX at all,
in case users are intending them to indicate regular text, in possible
ignorance of LaTeX. Instead, we want double-dollar-signs to be used to
indicate Inline Mode, and triple-dollar-signs to be used to indicate
Display (Block) Mode. A test has been added for this configuration.

[CH-25113](https://app.clubhouse.io/gradescope/story/25113/remark-math-supports-for-block-latex)

### CODE REVIEW

File: packages/remark-math/inline.js [hunk 1]
```diff
@@ -76,15 +128,15 @@ module.exports = function inlinePlugin (opts) {
   // Inject inlineTokenizer
   const inlineTokenizers = Parser.prototype.inlineTokenizers
   const inlineMethods = Parser.prototype.inlineMethods
-  inlineTokenizers.math = inlineTokenizer
-  inlineMethods.splice(inlineMethods.indexOf('text'), 0, 'math')
+  inlineTokenizers[NODE_TYPES.MATH_DISPLAY] = inlineTokenizer
```
Comment (ibrahima) [inline packages/remark-math/inline.js]: This part actually confuses me because it feels like the key should be `inlineMath`, or perhaps the original key being `math` had nothing to do with display-mode math? In general I am not sure whether defining this constant makes sense given that it's not being used (exclusively) for display-mode math nodes.

To me it just seems like the original author's intent was, let's add a math tokenizer to the inline tokenizers, and let's add a math tokenizer to the block tokenizer, but this doesn't necessarily relate to the node type. I think it just happens to have the same name. It seems like the inline parser doesn't ever emit `math` nodes by default, and with this PR we are allowing it to do so, but the default config also does not emit `math` nodes, it will only ever emit `inlineMath` nodes. Looking at some of the `remark-parse` code, not all of the tokenizers necessarily emit nodes of the type named by the tokenizer (e.g. [escape](https://github.com/remarkjs/remark/blob/remark-parse%405.0.0/packages/remark-parse/lib/tokenize/escape.js)).

---

TL;DR let's undo this change and uncouple these things, because they're not actually referring to the same thing. We can also get rid of the constant I think, because we don't it here.

File: packages/remark-math/inline.js [hunk 2]
```diff
@@ -2,72 +2,124 @@ function locator (value, fromIndex) {
...
+  MATH_DISPLAY: 'math',
+  MATH_INLINE: 'inlineMath'
+}
+
+const defaultModes = {
+  mathInlineDouble: {
+    nodeType: NODE_TYPES.MATH_INLINE,
+    left: /\$\$/,
+    right: /\$\$/,
+    matchInclude: [/\\\$/, /[^$]/],
```
Comment (ibrahima) [inline packages/remark-math/inline.js]: I am not totally sure if this needs to be a parameter, since it's the same for both current modes, and it feels slightly below the level of what you'd want to customize when you just want to change the delimiters.

It seems like this bit is just saying, in between two delimiters match anything that's not a dollar sign, and allow escaped dollar signs through as well. That doesn't seem like something that needs to change based on what delimiters are used, although I do like splitting it up into separate variables so that it's clearer what the regex is doing.

File: packages/remark-math/inline.js [hunk 3]
```diff
@@ -2,72 +2,124 @@ function locator (value, fromIndex) {
...
         value: value.slice(0, -2) + '$'
       })
     }
 
-    if (match) {
-      /* istanbul ignore if - never used (yet) */
-      if (silent) {
-        return true
-      }
+    if (!matchData) {
```
Comment (ibrahima) [inline packages/remark-math/inline.js:97]: Nice, I like that using this guard clause reduces the indentation level of everything that cares about whether there is a match.

File: packages/remark-math/inline.js [hunk 4]
```diff
@@ -2,72 +2,124 @@ function locator (value, fromIndex) {
...
+        value: toEat
       })
     }
+
+    const mode = modes[matchData.modeName]
+    return eat(fullMatch)({
+      type: mode.nodeType,
+      value: matchData.content,
+      data: {
+        hName: mode.tagName,
```
Comment (ibrahima) [inline packages/remark-math/inline.js]: Given that `div`s are not allowed to be inline, should we just leave the tagName hardcoded to `'span'`? It's funny because we're not even really using this `tagName` in the end, but I think in general it doesn't make sense to allow `div`s here so perhaps we can reduce the API surface by removing this parameter.

File: packages/remark-math/inline.js [hunk 5]
```diff
@@ -76,15 +128,15 @@ module.exports = function inlinePlugin (opts) {
...
+  inlineTokenizers[NODE_TYPES.MATH_DISPLAY] = inlineTokenizer
+  inlineMethods.splice(inlineMethods.indexOf(NODE_TYPES.TEXT), 0, NODE_TYPES.MATH_DISPLAY)
 
   const Compiler = this.Compiler
 
   // Stringify for math inline
   if (Compiler != null) {
     const visitors = Compiler.prototype.visitors
-    visitors.inlineMath = function (node) {
+    visitors[NODE_TYPES.MATH_INLINE] = function (node) {
```
Comment (ibrahima) [inline packages/remark-math/inline.js]: re: the thing above, this is actually fine because it is indeed referring to the node type.I think technically we don't use any of this functionality, and allowing the user to customize the delimiters will mean that this may not be consistent with the parser anymore. I think it already is inconsistent in the case that we've been using it with the disable single dollar signs thing so it's not a regression in any case.

File: packages/remark-math/inline.js [hunk 6]
```diff
@@ -2,72 +2,124 @@ function locator (value, fromIndex) {
   return value.indexOf('$', fromIndex)
```
Comment (ibrahima) [inline packages/remark-math/inline.js:2]: Not directly relevant to this PR, but I think it's worth noting that with the way this [locator][locator] thing works, if we wanted to do something like allow using `\[`/`\]` as delimiters as KaTeX does, this would prevent that from working. So we can't just pass in arbitrary left and right delimiters, they have to at least start with dollar signs.

[locator]: https://github.com/remarkjs/remark/blob/remark-parse%405.0.0/packages/remark-parse/readme.md#tokenizerlocatorvalue-fromindex

**Review by ibrahima — CHANGES_REQUESTED** [review]
Overall great work! Your changes make things much easier to understand. I do think that we can reduce the API surface a bit though, I think some parameters are not actually necessary. I like to compare to [KaTeX autorender's API](https://katex.org/docs/autorender.html#api) to see an example of a similar API that we might want to take inspiration from.

Thanks for adding test cases, they are very helpful!

File: packages/remark-math/inline.js (hunk 2 above)
Comment (rachelthompson) [reply packages/remark-math/inline.js]: I think I'd like to keep this as-is, if that's OK with you. What is an allowable internal match is related to what delimiters are used. I think that the fact that it's the same between the two expressions is somewhat a limitation of regular-expression-based parsing, and also the difficulties of deriving an inverse of a particular string. While this perhaps isn't ideal, it's the solution that seemed to be the best attainable ground to me, and it's (relatively) readable, and it works.

File: packages/remark-math/inline.js (hunk 5 above)
Comment (rachelthompson) [reply packages/remark-math/inline.js]: I'm not 100% clear - are you saying that it's OK to leave this line as-is?

Comment (ibrahima) [reply packages/remark-math/inline.js]: Sorry for not being clear, yup, this line is good to go. Just wanted to draw a contrast between this and the other `NODE_TYPES` usage.

Sometimes I have a tendency to ramble :/

File: packages/remark-math/inline.js (hunk 2 above)
Comment (ibrahima) [reply packages/remark-math/inline.js]: I thought that too, but I'm not so sure if that's the reason - you just can't have a dollar sign by itself in KaTeX even if you aren't using dollar sign delimiters. E.g. the demo on https://katex.org/ doesn't use any delimiters to denote latex, it's just parsing the whole input field. But you still get an error if you put in a dollar sign.

Playing around with some examples (trying to see what happens if we replace the middle with just `.+`:

https://regexr.com/4uos3
https://regexr.com/4uos9

Ah I kinda see, I guess the purpose of this might just be to make it so that it ends the match at the first dollar sign if one exists, but in the case of double dollar signs that doesn't make sense.

Double dollar signs:

https://regexr.com/4uosl
https://regexr.com/4uoso

Hmm, given that, it does kinda makes sense to change the middle part for double dollar signs, but it's weird to me that they did not do that for remark-math's existing double dollar sign regex, and we also have not done that here. I think the way it's set up it might not do multi-line matching so we wouldn't have that weirdness of going across lines, but it definitely would not render LaTeX there unless you escape the dollar sign.

Edit:

> I think that the fact that it's the same between the two expressions is somewhat a limitation of regular-expression-based parsing, and also the difficulties of deriving an inverse of a particular string

Yeah I think you're right, maybe that's the only reason they're the same. Or maybe no one realized it should be different.

---

At a higher level though, I feel like this is something that a consumer of this API wouldn't necessarily want to think about. At the least having some defaults would be nice IMO. I also feel that it would be nice if we could pass in strings for the left and right delimiters rather than regexes, but that's a minor thing. Again, I think [KaTeX auto-render's API](https://katex.org/docs/autorender.html#api) is nice and simple with the right level of abstraction.

---

Re: passing in strings, we could potentially do something like this: https://stackoverflow.com/questions/3446170/escape-string-for-use-in-javascript-regex but it is kinda annoying so it seems fine to keep the regex parameters.

**Comment by ibrahima** [comment]
One thing to keep in mind - I keep bringing up KaTeX's auto-render, partly because I think it's a good API, but it's also the thing we're trying to keep compatibility with, since that's what we've been using on other parts of the site. So it would be unfortunate if the parsing strategy we use is subtly different from that. I fear that that might happen regardless because they're not using regexes: https://github.com/KaTeX/KaTeX/tree/master/contrib/auto-render, but it'd be nice to try and make sure we handle edge cases in the same way. One way to test this would be to test some edge cases in other parts of Gradescope (e.g. rubric items) and see how it handles them.

It's not the end of the world if we don't have the exact same edge case handling as them, because those are edge cases and we can inform users of what they need to do to fix it (probably just escape their dollar signs or something like that).

File: packages/remark-math/inline.js [hunk 7]
```diff
@@ -2,72 +2,121 @@ function locator (value, fromIndex) {
...
-const INLINE_MATH_DOUBLE = /^\$\$((?:\\\$|[^$])+)\$\$/
+const ESCAPED_INLINE_MATH = /^\\\$/ // starts with \$
+
+const NODE_TYPES = {
+  TEXT: 'text',
+  MATH_INLINE: 'inlineMath'
+}
+
+const defaultModes = {
+  mathInlineDouble: {
```
Comment (ibrahima) [inline packages/remark-math/inline.js]: Oh yeah, I'm about to leave a similar comment on the other PR, but I think `inlineMathDouble` and `inlineMath` are a little more readable than this ordering, because it feels more like the regular order that you would say these words in a sentence. Also that's the order that the original regex constants used.

File: packages/remark-math/inline.js [hunk 8]
```diff
@@ -6,35 +6,46 @@ const ESCAPED_INLINE_MATH = /^\\\$/ // starts with \$
 
 const NODE_TYPES = {
   TEXT: 'text',
-  MATH_INLINE: 'inlineMath'
+  INLINE_MATH: 'inlineMath'
+}
+
+const includeNonDollarsAndEscaped = ['\\$', /[^$]/]
+
+function escapeRegExp (str) {
```
Comment (ibrahima) [inline packages/remark-math/inline.js:15]: Cool! I like this. Let's leave a note here about where it came from so that it doesn't look like voodoo:

```suggestion
// Taken from https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions#Escaping
function escapeRegExp (str) {
```

File: packages/remark-math/inline.js [hunk 9]
```diff
@@ -6,35 +6,46 @@ const ESCAPED_INLINE_MATH = /^\\\$/ // starts with \$
...
+  return str.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')
```
Comment (ibrahima) [inline packages/remark-math/inline.js]:
```suggestion
  return str.replace(/[.*+\-?^${}()|[\]\\]/g, '\\$&');
```

Tweaking this to match the MDN docs instead of the StackOverflow answer, because it seems slightly improved since it was copied to SO:

https://wiki.developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions$compare?locale=en-US&to=1607265&from=1605149

File: packages/remark-math/inline.js [hunk 10]
```diff
@@ -6,35 +6,46 @@ const ESCAPED_INLINE_MATH = /^\\\$/ // starts with \$
 
 const NODE_TYPES = {
   TEXT: 'text',
-  MATH_INLINE: 'inlineMath'
+  INLINE_MATH: 'inlineMath'
+}
+
+const includeNonDollarsAndEscaped = ['\\$', /[^$]/]
```
Comment (ibrahima) [inline packages/remark-math/inline.js:12]: Yay this is nice!

**Review by ibrahima — APPROVED** [review]
LGTM! I love the new API. Just gonna make one minor tweak to the regexp escaping to match MDN but otherwise this looks good!
